File: Android/FitIn_v1/app/src/main/java/com/example/fitin_v1/Util/AuthInterceptor.java
package com.example.fitin_v1.Util;

import android.util.Log;

import androidx.annotation.NonNull;

import com.example.fitin_v1.MyApp;
import com.example.fitin_v1.dto.TokenDto;
import com.example.fitin_v1.dto.TokenRequestDto;
import com.example.fitin_v1.remote.api.ReIssue;
import com.example.fitin_v1.remote.singleton.RetrofitBuilder;

import java.io.IOException;

import okhttp3.Interceptor;
import okhttp3.Request;
import okhttp3.Response;

public class AuthInterceptor implements Interceptor {


    @NonNull
    @Override
    public Response intercept(@NonNull Chain chain) throws IOException {
        ReIssue reIssue = RetrofitBuilder.getRetrofit().create(ReIssue.class);
        Utils.init(MyApp.getContext());
        String at = Utils.getAccessToken("none");
        String rt = Utils.getRefreshToken("nein");

        Request original = chain.request().newBuilder().header("Authorization", "Bearer " + at).build();
        Response response = chain.proceed(original);

        if (response.code() == 401) {

                TokenRequestDto tokenRequestDto = new TokenRequestDto(at, rt);
                retrofit2.Response<TokenDto> token = reIssue.getReIssue(tokenRequestDto).execute();
                if (token.isSuccessful()) {
                    Utils.setAccessToken(token.body().getAccessToken());
                    Utils.setRefreshToken(token.body().getRefreshToken());
//                    Log.e("Interceptor", "At:" + Utils.getAccessToken("nein"));
                    return chain.proceed(original.newBuilder().header("Authorization", "Bearer " + token.body().getAccessToken()).build());
                }


                return chain.proceed(original.newBuilder().header("Authorization", "Bearer " + Utils.getAccessToken("nein")).build());




        }
//        Log.e("Code What?", String.valueOf(response.code()));
        return response;
    }
}

File: Android/FitIn_Kotlin/app/src/main/java/com/example/fitin_kotlin/network/AuthInterceptor.kt
package com.example.fitin_kotlin.network

import com.example.fitin_kotlin.data.local.EncryptedSharedPreferenceController
import com.example.fitin_kotlin.data.model.network.request.RequestTokenReissue
import com.example.fitin_kotlin.data.model.network.response.ResponseToken
import com.example.fitin_kotlin.data.remote.api.UserService
import com.example.fitin_kotlin.data.repository.UserRepository
import com.example.fitin_kotlin.di.NetworkModule
import dagger.Lazy
import kotlinx.coroutines.*
import okhttp3.Interceptor
import okhttp3.Request
import okhttp3.Response
import javax.inject.Inject
import javax.inject.Qualifier
import javax.inject.Singleton

//@Qualifier
//@Retention(AnnotationRetention.BINARY)
//annotation class AuthInterceptorOkHttpClient

@Singleton
class AuthInterceptor @Inject constructor(private val prefs: EncryptedSharedPreferenceController, private val repository: Lazy<UserRepository>) : Interceptor{
    override fun intercept(chain: Interceptor.Chain): Response {

        val accessToken: String? = prefs.getAccessToken()
        val refreshToken: String? = prefs.getRefreshToken()

        val original: Request = chain.request().newBuilder().header("Authorization",
            "Bearer $accessToken"
        ).build()
        var response: Response = chain.proceed(original)

        if (response.code == 401 && response.body!!.equals("null")) {
            runBlocking {
                val requestTokenReissue = RequestTokenReissue(accessToken.toString(), refreshToken.toString())
                val token: retrofit2.Response<ResponseToken> = repository.get().postReIssue(requestTokenReissue)
                withContext(Dispatchers.Main) {
                    if(token.isSuccessful) {
                        prefs.setAccessToken(token.body()!!.accessToken)
                        prefs.setRefreshToken(token.body()!!.refreshToken)
                        response = chain.proceed(original.newBuilder().header("Authorization", "Bearer " + token.body()!!.accessToken).build())
                    }
                    response = chain.proceed(original.newBuilder().header("Authorization", "Bearer " + prefs.getAccessToken()).build())
                }
            }
        }
        return response
    }
}